src/Microsoft.DocAsCode.MarkdownLite/Matchers/Matcher.cs

```diff
@@ -0,0 +1,154 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+namespace Microsoft.DocAsCode.MarkdownLite.Matchers
+{
+    using System;
+
+    public abstract class Matcher
+    {
+        public const int NotMatch = -1;
+
+        private static readonly AnyCharMatcher AnyCharMatcher = new AnyCharMatcher();
+        private static readonly EosMatcher EosMatcher = new EosMatcher();
+
+        /// <summary>
+        /// 
+        /// </summary>
+        /// <param name="text"></param>
+        /// <param name="startIndex"></param>
+        /// <returns>Char count of match, <c>-1</c> is not match.</returns>
+        public abstract int Match(MatchContent content);
+
+        public static Matcher Char(char ch)
+        {
+            return new CharMatcher(ch);
+        }
+
+        public static Matcher AnyChar() => AnyCharMatcher;
+
+        public static Matcher AnyCharIn(params char[] ch)
+        {
+            if (ch == null)
+            {
+                throw new ArgumentNullException(nameof(ch));
+            }
+            var array = (char[])ch.Clone();
+            Array.Sort(array);
+            return new AnyCharInMatcher(ch);
+        }
+
+        public static Matcher AnyCharInRange(char start, char end)
+        {
+            if (start >= end)
+            {
+                throw new ArgumentException(nameof(end), $"Should great than {start.ToString()}");
+            }
+            return new AnyCharInRangeMatcher(start, end);
+        }
+
+        public static Matcher AnyCharNotIn(params char[] ch)
+        {
+            if (ch == null)
+            {
+                throw new ArgumentNullException(nameof(ch));
+            }
+            var array = (char[])ch.Clone();
+            Array.Sort(array);
+            return new AnyCharNotInMatcher(ch);
+        }
+
+        public static Matcher String(string text)
+        {
+            if (text == null)
+            {
+                throw new ArgumentNullException(nameof(text));
+            }
+            if (text.Length == 0)
+            {
+                throw new ArgumentException("Cannot be empty.", nameof(text));
+            }
+            return new StringMatcher(text);
+        }
+
+        public static Matcher Eos() => EosMatcher;
+
+        public static Matcher Maybe(Matcher matcher) =>
+            Repeat(matcher, 0, 1);
+
+        public static Matcher Repeat(Matcher matcher, int minOccur) =>
+            Repeat(matcher, minOccur, int.MaxValue);
```

Comment:
like regex: [^\n]*

---
In reply to: [104580437](https://github.com/dotnet/docfx/pull/1388#discussion_r104580437) [](ancestors = 104580437)